apps/omg_watcher_info/lib/omg_watcher_info/transaction.ex

```diff
@@ -0,0 +1,252 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.WatcherInfo.Transaction do
+  @moduledoc """
+  Module create transaction from selected utxos and order.
+  """
+
+  alias OMG.Crypto
+  alias OMG.State.Transaction
+  alias OMG.TypedDataHash
+  alias OMG.WatcherInfo.DB
+  alias OMG.WatcherInfo.UtxoSelection
+
+  require Transaction.Payment
+
+  @empty_metadata <<0::256>>
+  @max_outputs Transaction.Payment.max_outputs()
+  @merge_fee 0
+
+  @type create_t() ::
+          {:ok, nonempty_list(transaction_t())}
+          | {:error, :too_many_outputs}
+          | {:error, :empty_transaction}
+
+  @type create_typed_data_t() ::
+          {:ok, nonempty_list(transaction_with_typed_data_t())}
+          | {:error, :too_many_outputs}
+          | {:error, :empty_transaction}
+
+  @type fee_t() :: %{
+          currency: UtxoSelection.currency_t(),
+          amount: non_neg_integer()
+        }
+  @type payment_t() :: %{
+          owner: Crypto.address_t() | nil,
+          currency: UtxoSelection.currency_t(),
+          amount: pos_integer()
+        }
+
+  @type transaction_t() :: %{
+          inputs: nonempty_list(%DB.TxOutput{}),
+          outputs: nonempty_list(payment_t()),
+          fee: fee_t(),
+          txbytes: Transaction.tx_bytes() | nil,
+          metadata: Transaction.metadata(),
+          sign_hash: Crypto.hash_t() | nil
+        }
+
+  @type transaction_with_typed_data_t() :: %{
+          inputs: nonempty_list(%DB.TxOutput{}),
+          outputs: nonempty_list(payment_t()),
+          fee: fee_t(),
+          txbytes: Transaction.tx_bytes() | nil,
+          metadata: Transaction.metadata(),
+          sign_hash: Crypto.hash_t() | nil,
+          typed_data: TypedDataHash.Types.typedDataSignRequest_t()
+        }
+
+  @type order_t() :: %{
+          owner: Crypto.address_t(),
+          payments: list(payment_t()),
+          metadata: binary() | nil,
+          fee: fee_t()
+        }
+
+  @type utxos_map_t() :: %{UtxoSelection.currency_t() => UtxoSelection.utxo_list_t()}
+  @type inputs_t() :: {:ok, utxos_map_t()} | {:error, {:insufficient_funds, list(map())}}
+
+  @doc """
+  Given an `order`, finds spender's inputs sufficient to perform a payment.
+  If also provided with receiver's address, creates and encodes a transaction.
+  """
+  @spec select_inputs(utxos_map_t(), order_t()) :: inputs_t()
+  def select_inputs(utxos, %{payments: payments, fee: fee}) do
+    payments
+    |> UtxoSelection.needed_funds(fee)
```

Comment:
I think we could rename these functions to be more explicit on what they do. It's a bit confusing.
(I'm talking about `needed_funds`, `select_utxo`, `funds_sufficient`).
Or/and you could also add more comments in the code to tell exactly what it does.